cqc/pythonLib.py

```diff
@@ -2379,3 +2471,352 @@ def getTime(self, block=True):
             return otherHdr.datetime
         except AttributeError:
             return None
+
+
+class CQCToFile(CQCHandler):
+    """Handler to be used when writing the CQC commands to a file."""
+
+    def __init__(self, filename='CQC_File', pend_messages=False,
+                 overwrite=False):
+
+        # Call init of CQCHandler
+        super().__init__(pend_messages=pend_messages)
+
+        self.name = "CQCToFile"
+
+        self.next_qubitID = 0
+
+        # Set path of file to write to
+        script_dir = sys.path[0]
+        self.filename = os.path.join(script_dir, filename)
+        self.filenameb = self.filename + "binary"
+
+        # Check if file exists
+        if overwrite:
+            # Remove file if we can overwrite
+            try:
+                os.remove(self.filename)
+            except FileNotFoundError:
+                pass
+            
+            try:
+                os.remove(self.filenameb)
+            except FileNotFoundError:
+                pass
+        else:
+            if not (os.path.isfile(self.filename) 
+                    or os.path.isfile(self.filename + "binary")):
+                pass
+            else:
+                # Append number to filename if can't overwrite
+                num = 0
+                while True:
+                    if (os.path.isfile(self.filename + str(num)) 
+                            or os.path.isfile(self.filename + str(num) 
+                                              + "binary")):
+                        num += 1
+                    else:
+                        self.filename = self.filename + str(num)
+                        self.filenameb = self.filename + "binary"
+                        break 
+
+        # Don't want notify when writing to file
+        self.notify = False
+
+    def commit(self, msg):
+        """Write a message to file.
+
+        Message is written as the bytes turned into a string.
+        """
+
+        with open(self.filename, 'a') as f:
+            f.write(str(msg) + '\n')
+
+        with open(self.filenameb, 'ab') as f:
+            f.write(msg)
+
+    def new_qubitID(self):
+        """Provice new qubit ID."""
+
+        qubitID = self.next_qubitID
+
+        self.next_qubitID += 1
+
+        return qubitID
+
+    def close(self, release_qubits=True):
+        """Handle exiting context."""
+
+        # Flush all remaining commands
+        if self._pending_headers:
+            self.flush()
+
+        if release_qubits and self.active_qubits:
+
+            msg = self.construct_release(self.active_qubits[:], notify=True, 
+                                         block=True, action=False)
+            self.commit(msg)
+
+    def createEPR(self, name, remote_appID=0, notify=True, block=True):
```

Comment:
Same comment as above, I think we don't need to have this logic in both of the subclasses. Or maybe we do but I cannot see why at this point

Reply:
Do you mean the whole `createEPR` method?

Reply:
Yeah, for `createEPR`, `recvEPR`, `sendQubit`, `recvQubit` and `flush_factory` the code is mostly the same for the two classes. It would be better to avoid this.